### PR TITLE
Backend dependency updates

### PR DESCRIPTION
**Pull request by stevejalim**
## What's changing?

* Update compile-requirements script to drop compiled deps before recompiling new ones, meaning we get updates to subdeps more often

* Dependency bumps

## Testing

Integration tests: https://github.com/mozilla/bedrock/actions/runs/7386413501

### CODE REVIEW

File: bin/compile-requirements.sh [hunk 1]
```diff
@@ -12,8 +12,11 @@ export CUSTOM_COMPILE_COMMAND="$ make compile-requirements"
 # We need this installed, but we don't want it to live in the main requirements
 # We will need to periodically review this pinning
 
-pip install -U pip==23.1.2
-pip install pip-tools==6.13.0
+pip install -U pip==23.3.2
+pip install pip-tools==7.3.0
+
+# Drop the compiled reqs files, to help us pick up automatic subdep updates, too
+rm -f requirements/*.txt
```
Comment (stevejalim) [inline bin/compile-requirements.sh:19]: @robhudson Am I overlooking a benefit of keeping the previously compiled reqs files? With them present, pip-compile definitely picks up fewer changes, very commonly only the top-level changes and no natural/allowed subdep changes

Comment (robhudson) [reply bin/compile-requirements.sh:19]: Not sure, but I like the idea of picking up more subdep changes so +1 if this helps make that happen.

File: requirements/dev.in [hunk 2]
```diff
@@ -1,25 +1,25 @@
 -r prod.txt
 
-black==23.11.0
+black==23.12.1
```
Comment (robhudson) [inline requirements/dev.in:3]: I have a branch where I moved to Ruff. Should I dust it off and push it up for review?

**Review by robhudson — APPROVED** [review]
🏞️

**Pull request merged**